llap-tez/src/java/org/apache/hadoop/hive/llap/tezplugins/metrics/BlacklistingLlapMetricsListener.java

```diff
@@ -0,0 +1,200 @@
+    this.emptyExecutorsThreshold =
+        HiveConf.getFloatVar(conf,
+            HiveConf.ConfVars.LLAP_TASK_SCHEDULER_BLACKLISTING_METRICS_LISTENER_EMPTY_EXECUTORS);
+    this.maxBlacklistedNodes =
+        HiveConf.getIntVar(conf, HiveConf.ConfVars.LLAP_TASK_SCHEDULER_BLACKLISTING_METRICS_LISTENER_MAX_LISTED_NODES);
+
+    Preconditions.checkArgument(minServedTasksNumber > 0,
+        "Minimum served tasks should be greater than 0");
+    Preconditions.checkArgument(minConfigChangeDelayMs > 0,
+        "Minimum config change delay should be greater than 0");
+    Preconditions.checkArgument(timeThreshold > 1.0f,
+        "The time threshold should be greater than 1");
+    Preconditions.checkArgument(maxBlacklistedNodes > 0,
+        "The maximum number of blacklisted node should be greater than 1");
+
+    LOG.debug("BlacklistingLlapMetricsListener initialized with " +
+                  "minServedTasksNumber={}, " +
+                  "minConfigChangeDelayMs={}, " +
+                  "timeThreshold={}, " +
+                  "emptyExecutorsThreshold={}, " +
+                  "maxBlacklistedNodes={}",
+        minServedTasksNumber, minConfigChangeDelayMs, timeThreshold, emptyExecutorsThreshold, maxBlacklistedNodes);
+  }
+
+  @Override
+  public void init(Configuration conf, LlapRegistryService registry) {
+    init(conf, registry, LlapManagementProtocolClientImplFactory.basicInstance(conf));
+  }
+
+  @Override
+  public void newDaemonMetrics(String workerIdentity, LlapMetricsCollector.LlapMetrics newMetrics) {
+    // no op
+  }
+
+  @Override
+  public void newClusterMetrics(Map<String, LlapMetricsCollector.LlapMetrics> newMetrics) {
+    long sumAverageTime = 0;
+    long sumEmptyExecutors = 0;
+    long maxAverageTime = 0;
+    long maxAverageTimeEmptyExecutors = 0;
+    long maxAverageTimeMaxExecutors = 0;
+    long workerNum = 0;
+    int blacklistedNodes = 0;
+    String maxAverageTimeIdentity = null;
+    for (String workerIdentity : newMetrics.keySet()) {
+      Map<String, Long> metrics = newMetrics.get(workerIdentity).getMetrics();
+      long requestHandled = metrics.get(LlapDaemonExecutorInfo.ExecutorTotalRequestsHandled.name());
+      long averageTime = metrics.get(LlapDaemonExecutorInfo.AverageResponseTime.name());
+      long emptyExecutor =
+          metrics.get(LlapDaemonExecutorInfo.ExecutorNumExecutorsAvailableAverage.name());
+      long maxExecutors = metrics.get(LlapDaemonExecutorInfo.ExecutorNumExecutorsPerInstance.name());
+
+      LOG.debug("Checking node {} with data: " +
+                    "requestHandled={}, " +
+                    "averageTime={}, " +
+                    "emptyExecutors={}, " +
+                    "maxExecutors={}",
+          workerIdentity, requestHandled, averageTime, emptyExecutor, maxExecutors);
+
+      if (maxExecutors == 0) {
+        blacklistedNodes++;
+        if (blacklistedNodes >= this.maxBlacklistedNodes) {
```

Comment:
Same here: How can we get into a situation where we exceed the maxBlacklistedNodes?
I would assume that we can reach but never exceed it. (which would make the LOG message misleading).

Reply:
I think/hope that we do not end up with more blacklisted nodes than we planned, but more robust this way - see my bug which would have allowed to blacklist more nodes.
Also the API is available for everyone. So they might be able to call it outside this scope.

Reply:
And changed the log message and level..